crates/turborepo-lib/src/shim.rs

```diff
@@ -176,50 +177,237 @@ struct PackageJson {
+        {
             #[cfg(windows)]
             {
-                "turbo.cmd"
+                "turbo.exe"
             }
             #[cfg(not(windows))]
             {
                 "turbo"
             }
-        });
-
-        if !local_turbo_path.exists() {
-            debug!(
-                "No local turbo binary found at: {}",
-                local_turbo_path.display()
-            );
-            return None;
         }
+    }
+
+    #[allow(dead_code)]
+    pub fn version() -> &'static str {
+        include_str!("../../../version.txt")
+            .lines()
+            .next()
+            .expect("Failed to read version from version.txt")
+    }
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct LocalTurboState {
+    bin_path: PathBuf,
+    version: String,
+}
+
+impl LocalTurboState {
+    // Hoisted strategy:
+    // - `npm install`
+    // - `yarn`
+    // - `yarn install --flat`
+    // - berry (nodeLinker: "node-modules")
+    //
+    // This also supports people directly depending upon the platform version.
+    fn generate_hoisted_path(root_path: &Path) -> Option<PathBuf> {
+        Some(root_path.join("node_modules"))
+    }
+
+    // Nested strategy:
+    // - `npm install --install-strategy=shallow` (`npm install --global-style`)
+    // - `npm install --install-strategy=nested` (`npm install --legacy-bundling`)
+    // - berry (nodeLinker: "pnpm")
+    fn generate_nested_path(root_path: &Path) -> Option<PathBuf> {
+        Some(
+            root_path
+                .join("node_modules")
+                .join("turbo")
+                .join("node_modules"),
+        )
+    }
 
-        let local_turbo_package_path = repo_root
-            .join("node_modules")
-            .join("turbo")
+    // Linked strategy:
+    // - `pnpm install`
+    // - `npm install --install-strategy=linked`
+    fn generate_linked_path(root_path: &Path) -> Option<PathBuf> {
+        fs_canonicalize(root_path.join("node_modules").join("turbo").join("..")).ok()
+    }
+
+    // The unplugged directory doesn't have a fixed path.
+    fn get_unplugged_base_path(root_path: &Path) -> PathBuf {
+        let yarn_rc_filename =
+            env::var_os("YARN_RC_FILENAME").unwrap_or_else(|| OsString::from(".yarnrc.yml"));
+        let yarn_rc_filepath = root_path.join(yarn_rc_filename);
+
+        let yarn_rc_yaml_string = fs::read_to_string(yarn_rc_filepath).unwrap_or_default();
+        let yarn_rc: YarnRc = serde_yaml::from_str(&yarn_rc_yaml_string).unwrap_or_default();
+
+        root_path.join(yarn_rc.pnp_unplugged_folder)
+    }
+
+    // Unplugged strategy:
+    // - berry 2.1+
+    fn generate_unplugged_path(root_path: &Path) -> Option<PathBuf> {
+        let platform_package_name = TurboState::platform_package_name();
+        let unplugged_base_path = Self::get_unplugged_base_path(root_path);
+
+        unplugged_base_path
+            .read_dir()
+            .ok()
+            .and_then(|mut read_dir| {
```

Comment:
This works great, but just FYI, you can also do:

```
let mut read_dir = unplugged_base_path.read_dir().ok()?;
```

Because `?` works with both `Option` and `Result`

Reply:
It made the chaining less elegant when I had it that way. Since my target outcome is an `Option<PathBuf>` `?` pulls me all the way out, and `and_then` doesn't do the nice resolution for the chain.

I was using `?` before changing the API to `Option<PathBuf>`, though!